posthog/caching/insights_api.py

```diff
@@ -1,46 +1,95 @@
 
-    if (hasattr(filter, "interval") and filter.interval == "hour") or (delta_days is not None and delta_days <= 7):
-        refresh_frequency = timedelta(minutes=3)
+    refresh_frequency = BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL
+    if getattr(filter, "interval", None) == "hour" or (delta_days is not None and delta_days <= 7):
+        if not is_shared:  # The interval is always longer for shared insights/dashboard
+            refresh_frequency = REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+    elif is_shared:  # The interval is always longer for shared insights/dashboard
+        refresh_frequency = INCREASED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+
+    refresh_insight_now = False
+    # Only refresh if the user has requested a refresh or if we're on a shared dashboard/insight
+    # (those have no explicit way of reloading)
+    if refresh_requested_by_client(request) or is_shared:
+        are_there_cached_results = caching_state is not None and caching_state.last_refresh is not None
+        is_data_cached_but_stale = are_there_cached_results and (
+            cast(InsightCachingState, caching_state).last_refresh + refresh_frequency <= now
+        )
+        refresh_insight_now = (not are_there_cached_results) or is_data_cached_but_stale
 
-    if len(caching_state) == 0 or caching_state[0].last_refresh is None:
-        return True, refresh_frequency
+    # Check if the refresh is might already be running for this cache key - if so, let's wait until that's done
+    if refresh_insight_now and caching_state is not None and caching_state.last_refresh_queued_at is not None:
+        while (
+            # Refresh must have been queued recently enough that the query might still be running
+            caching_state.last_refresh_queued_at > now - timedelta(seconds=CLICKHOUSE_MAX_EXECUTION_TIME)
+            # Also, refreshing must have either never finished or last finished before it was queued now
+            and (
+                caching_state.last_refresh is None or caching_state.last_refresh < caching_state.last_refresh_queued_at
+            )
+        ):
+            # This looks a bit ugly because it blocks the thread, but it's BETTER than running the already-running query
+            # from this thread concurrently. Either way we have to wait for results, this way just hammers the DB less
+            sleep(3)
+            caching_state.refresh_from_db()
+            if (
+                caching_state.last_refresh is not None
+                and caching_state.last_refresh >= caching_state.last_refresh_queued_at
+            ):
+                refresh_insight_now = False
+                break  # Refreshed successfully!
+        # Otherwise we're sure the refresh isn't running at the moment - either it's not been queued or it's timed out
+        # (barring the occasional race condition related to fetching state from PG, but that much uncertainty is okay)
```

Comment:
Good point @mariusandra, we do already have `last_refresh_queued_at`. What do you think about this `sleep`y solution?

Reply:
This is definitely better, though my worry now is not hammering the ClickHouse DB, but exhausting available nginx connection slots or django sync workers. These are also finite, if put behind a firehose of traffic. Under a nightmare "Yahoo News Embed" scenario, no real traffic will reach app anymore. 

What could work here? Server sent events? Async polling on the client side? 🤔

Reply:
Polling would be relatively simple, but still much more involved than this. Looking at the [data from Cloud US Grafana](http://grafana-prod-us/explore?left=%7B%22datasource%22:%22PBFA97CFB590B2093%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22PBFA97CFB590B2093%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%28increase%28django_http_requests_total_by_view_transport_method_total%7Bview%3D~%5C%22%28posthog%5C%5C%5C%5C.api%5C%5C%5C%5C.sharing%5C%5C%5C%5C.SharingViewerPageViewSet%7Cproject_insights-trend%7Cproject_dashboards-detail%29%5C%22%7D%5B$__interval%5D%29%29%20by%20%28view%29%22,%22interval%22:%22240m%22,%22range%22:true,%22refId%22:%22A%22%7D%5D,%22range%22:%7B%22from%22:%22now-7d%22,%22to%22:%22now%22%7D%7D&orgId=1), it seems unlikely for shared insight/dashboard to block threads more than all the other ways of calculating insights. I think this should be fine.

Reply:
With calculated insights and personal API keys or cookies, there's a clear correlation between requests and users, and we're able to handle the load. 

Here, we're talking about potentially queueing 10k-100k requests by that many unique users in a few seconds, exhausting all available django and nginx workers... I've seen this happen before and it's not pretty. I don't feel comfortable signing off on this without someone from team infra ack'ing that this is not a Bad Idea TM.

Reply:
Fair enough, made it so that we don't wait on shared insights/dashboards. There are still some test failures where a test expects the results to never have been calculated, which seems odd. Chesterton's fence though, so I'd also rather not merge not knowing why those test want results to be null. Paul, Karl, and maybe Yakko are the people who might have some context on this… none of them available this week though. 😅

Reply:
`I'd also rather not merge not knowing why those test want results to be null`

There's definitely some behaviour change here.

IIRC if there is nothing in the cache we return no results (which is what those failing tests are checking for). We can return no results because we know that the front end will make an api call to ask for the results to be refreshed.

I've always assumed this is for speed. The rest of the information comes back so the dashboard page can be drawn and will show tiles with spinners, rather than the whole page being a spinner.

This doesn't work for shared dashboards which can't make an API call... 

I guess that this PR has changed things so that we wait for results when the dashboard is not shared.

The safer change might be to only change behaviour for shared dashboards. Check it fixes the issue and then tidy up...